pydantic/_internal/_fields.py

```diff
@@ -274,11 +274,12 @@ def collect_dataclass_fields(
 
     fields: dict[str, FieldInfo] = {}
     dataclass_fields: dict[str, dataclasses.Field] = cls.__dataclass_fields__
-    cls_localns = dict(vars(cls))  # this matches get_cls_type_hints_lenient, but all tests pass with `= None` instead
+    # this matches get_cls_type_hints_lenient, but all tests pass with `= None` instead
+    cls_localns = NsResolver(ImmutableNs(dict(vars(cls))))
 
     source_module = sys.modules.get(cls.__module__)
     if source_module is not None:
-        types_namespace = {**source_module.__dict__, **(types_namespace or {})}
+        types_namespace = NsResolver(ImmutableNs(source_module.__dict__), types_namespace)
```

Comment:
We should probably be careful with precedence here - in `dict` updates, the latter dicts override duplicate keys in previous ones. We should follow the same pattern in `NsResolver`.

Reply:
Added tests now for this: 
- latter overrides duplicate keys in previous ones
- "push" behaviour